overseer/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+	/// let spawner = executor::ThreadPool::new().unwrap();
+	/// let (overseer, _handler) = Overseer::new(
+	///     Box::new(ValidationSubsystem),
+	///     Box::new(CandidateBackingSubsystem),
+	///     spawner,
+	/// );
+	///
+	/// let timer = Delay::new(Duration::from_secs(1)).fuse();
+	///
+	/// let overseer_fut = overseer.run().fuse();
+	/// pin_mut!(timer);
+	/// pin_mut!(overseer_fut);
+	///
+	/// select! {
+	///     _ = overseer_fut => (),
+	///     _ = timer => (),
+	/// }
+	/// #
+	/// # }); }
+	/// ```
+	pub fn new(
+		validation: Box<dyn Subsystem<ValidationSubsystemMessage> + Send>,
+		candidate_backing: Box<dyn Subsystem<CandidateBackingSubsystemMessage> + Send>,
+		mut s: S,
+	) -> (Self, OverseerHandler) {
+		let (events_tx, events_rx) = mpsc::channel(CHANNEL_CAPACITY);
+
+		let handler = OverseerHandler {
+			events_tx: events_tx.clone(),
+		};
+
+		let mut running_subsystems_rx = StreamUnordered::new();
+		let mut running_subsystems = FuturesUnordered::new();
+
+		let validation_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			validation,
+		);
+
+		let candidate_backing_subsystem = Self::spawn(
+			&mut s,
+			&mut running_subsystems,
+			&mut running_subsystems_rx,
+			candidate_backing,
+		);
+
+		let this = Self {
+			validation_subsystem,
+			candidate_backing_subsystem,
+			s,
+			running_subsystems,
+			running_subsystems_rx,
+			events_rx,
+		};
+
+		(this, handler)
+	}
+
+	// Stop the overseer.
+	async fn stop(mut self) {
+		if let Some(ref mut s) = self.validation_subsystem.instance {
+			let _ = s.tx.send(FromOverseer::Signal(OverseerSignal::StopWork)).await;
+		}
+
+		if let Some(ref mut s) = self.candidate_backing_subsystem.instance {
+			let _ = s.tx.send(FromOverseer::Signal(OverseerSignal::StopWork)).await;
+		}
+
+		let mut stop_delay = Delay::new(Duration::from_secs(STOP_DELAY)).fuse();
+
+		loop {
+			select! {
+				_ = self.running_subsystems.next() => {
+					if self.running_subsystems.is_empty() {
+						break;
+					}
+				},
+				_ = stop_delay => break,
+				complete => break,
+			}
+		}
+	}
+
+	/// Run the `Overseer`.
+	pub async fn run(mut self) {
+		loop {
+			while let Poll::Ready(Some(msg)) = poll!(&mut self.events_rx.next()) {
+				match msg {
+					Event::MsgToSubsystem(msg) => {
+						self.route_message(msg).await;
+					}
+					Event::Stop => return self.stop().await,
+					_ => ()
+				}
+			}
+
+			while let Poll::Ready(Some((StreamYield::Item(msg), _))) = poll!(
+				&mut self.running_subsystems_rx.next()
+			) {
+				match msg {
+					ToOverseer::SubsystemMessage(msg) => self.route_message(msg).await,
+					ToOverseer::SpawnJob { s, res } => {
+						let s = self.spawn_job(s);
+
+						let _ = res.send(s);
+					}
+				}
+			}
+
+			// Some subsystem exited? It's time to panic.
+			if let Poll::Ready(Some(finished)) = poll!(self.running_subsystems.next()) {
+				panic!("Subsystem finished unexpectedly {:?}", finished);
```

Comment:
Any possibility for a more graceful (but still immediate) exit? `substrate-service` does something similar with `essential_tasks` whose failure must take down the entire node.

Reply:
changed to resolving a future with an error